### ISSUE TITLE
Argument was out of range of valid values.

### ISSUE THREAD

**Issue by isaacabraham** [mbraceproject/MBrace.Core#169]
A strange issue here. Create a clean file as follows: -

``` fsharp

/// fix package refs depending on your path....
#load @"..\packages\MBrace.Thespian\MBrace.Thespian.fsx"
#r @"..\packages\MBrace.Flow\lib\net45\MBrace.Flow.dll"
#r @"..\packages\Streams\lib\net45\Streams.dll"
#load @"dependencies.fsx"

open MBrace.Core
open MBrace.Flow
open MBrace.Library.Cloud
open MBrace.Thespian
open System

let cluster = ThespianCluster.InitOnCurrentMachine 1

[| 1 .. 10 |]
|> CloudFlow.OfArray
|> CloudFlow.map(fun n -> n * 2)
|> CloudFlow.toArray
|> cluster.Run
```

And have `dependencies.fsx` contain one of the two break repos from https://gist.github.com/isaacabraham/dbb2b77057e31990126019e41eb533be. The code above fails for me: -

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   at Mono.Collections.Generic.Collection`1.get_Item(Int32 index)
   at Mono.Cecil.ImportGenericContext.MethodParameter(String method, Int32 position)
   at Mono.Cecil.MetadataImporter.ImportGenericParameter(Type type, ImportGenericContext context)
   at Mono.Cecil.MetadataImporter.ImportTypeSpecification(Type type, ImportGenericContext context)
   at Mono.Cecil.MetadataImporter.ImportType(Type type, ImportGenericContext context, ImportGenericKind import_kind)
   at Mono.Cecil.MetadataImporter.ImportGenericInstance(Type type, ImportGenericContext context)
   at Mono.Cecil.MetadataImporter.ImportTypeSpecification(Type type, ImportGenericContext context)
   at Mono.Cecil.MetadataImporter.ImportType(Type type, ImportGenericContext context, ImportGenericKind import_kind)
   at Mono.Cecil.MetadataImporter.ImportMethodSpecification(MethodBase method, ImportGenericContext context)
   at Mono.Cecil.MetadataImporter.ImportMethod(MethodBase method, ImportGenericContext context, ImportGenericKind import_kind)
   at Mono.Cecil.ModuleDefinition.Import(MethodBase method, IGenericParameterProvider context)
   at Nessos.Vagabond.AssemblyParser.AssemblyParser.CreateReference(MethodBase method, MethodReference context)
   at Nessos.Vagabond.AssemblyParser.AssemblyParser.MapInstructions(MethodBase method, MethodDefinition method_definition)
   at Nessos.Vagabond.AssemblyParser.AssemblyParser.MapMethodBody(MethodBase method, MethodDefinition method_definition)
   at Nessos.Vagabond.AssemblyParser.AssemblyParser.MapMethod(TypeDefinition type_definition, MethodBase method)
   at Nessos.Vagabond.AssemblyParser.AssemblyParser.MapType(Type type, TypeDefinition declaringType)
   at Nessos.Vagabond.AssemblyParser.AssemblyParser.MapType(Type type, TypeDefinition declaringType)
   at Nessos.Vagabond.AssemblyParser.AssemblyParser.Map()
   at Nessos.Vagabond.AssemblyParser.parseDynamicAssemblySlice(AssemblyCompilerState state, Assembly assembly) in C:\\Users\\eirik\\Development\\nessos\\Vagabond\\src\\Vagabond\\SliceCompiler\\AssemblyParser.fs:line 182
   at Nessos.Vagabond.SliceCompiler.compileDynamicAssemblySlice(AssemblyCompilerState state, Assembly dynamicAssembly) in C:\\Users\\eirik\\Development\\nessos\\Vagabond\\src\\Vagabond\\SliceCompiler\\AssemblyCompiler.fs:line 71
   at Nessos.Vagabond.SliceCompiler.compileAssembly(AssemblyCompilerState state, Assembly assembly) in C:\\Users\\eirik\\Development\\nessos\\Vagabond\\src\\Vagabond\\SliceCompiler\\AssemblyCompiler.fs:line 124
   at Nessos.Vagabond.SliceCompiler.compileAssemblies@134-1.Invoke(Assembly assembly) in C:\\Users\\eirik\\Development\\nessos\\Vagabond\\src\\Vagabond\\SliceCompiler\\AssemblyCompiler.fs:line 137
   at Nessos.Vagabond.Utils.MailboxProcessor`1.PostAndReply[Msg,R](FSharpMailboxProcessor`1 m, FSharpFunc`2 msgB) in C:\\Users\\eirik\\Development\\nessos\\Vagabond\\src\\Vagabond\\Utils.fs:line 302
   at Nessos.Vagabond.VagabondManager.compile(Assembly[] assemblies) in C:\\Users\\eirik\\Development\\nessos\\Vagabond\\src\\Vagabond\\Vagabond.fs:line 34
   at Nessos.Vagabond.VagabondManager.ComputeObjectDependencies(Object graph, FSharpOption`1 permitCompilation, FSharpOption`1 includeNativeDependencies) in C:\\Users\\eirik\\Development\\nessos\\Vagabond\\src\\Vagabond\\Vagabond.fs:line 111
   at MBrace.Runtime.Components.StoreAssemblyManager.ComputeDependencies[T](T graph) in C:\\Users\\eirik\\Development\\mbrace\\MBrace.Core\\src\\MBrace.Runtime\\Components\\StoreAssemblyManager.fs:line 294
   at MBrace.Runtime.Components.StoreAssemblyManager.MBrace-Runtime-IAssemblyManager-ComputeDependencies(Object graph) in C:\\Users\\eirik\\Development\\mbrace\\MBrace.Core\\src\\MBrace.Runtime\\Components\\StoreAssemblyManager.fs:line 308
   at <StartupCode$MBrace-Runtime>.$MBraceClient.CreateProcessAsync@65.Invoke(Unit unitVar) in C:\\Users\\eirik\\Development\\mbrace\\MBrace.Core\\src\\MBrace.Runtime\\Runtime\\MBraceClient.fs:line 66
   at Microsoft.FSharp.Control.AsyncBuilderImpl.callA@851.Invoke(AsyncParams`1 args)
   at MBrace.Core.Internals.AsyncExtensions.Async.RunSync[T](FSharpAsync`1 workflow, FSharpOption`1 cancellationToken) in C:\\Users\\eirik\\Development\\mbrace\\MBrace.Core\\src\\MBrace.Core\\Utils\\AsyncExtensions.fs:line 101
```

break1-repro goes away if you remove the custom >>= operator. break2-repro goes away if you replace the call to Async.map in AsyncResult.map; I've included a "fix" in the gist (break2-solve.fs) which might give you some clues as to why it occurs.

I have no clue what is at fault here - MBrace.Core, MBrace.Thespain, Vagabond etc.

**Comment by eiriktsarpalis** [comment]
Hmm, I cannot reproduce this with the latest thespian (1.2.6) on my machine. But since Vagabond related-issues are sensitive to the way interactions are grouped together, perhaps the problem is triggered after a non-trivial succession of repl interactions. I tried a few combinations but couldn't get it to crash. Both sending the whole file at once and doing it line-by-line.

**Comment by isaacabraham** [comment]
Let me try again - I'll also check which versions of the dependencies that I have.

**Comment by isaacabraham** [comment]
Same issue here I'm afraid with a clean solution. Not sure what it is. Here's my lock file: -

```
REDIRECTS: ON
FRAMEWORK: NET452
NUGET
  remote: https://www.nuget.org/api/v2
    Argu (2.1)
    FsPickler (2.1)
    FsPickler.Json (2.1)
      FsPickler (2.1)
      Newtonsoft.Json (>= 6.0.5)
    MBrace.Azure (1.3)
      Argu (>= 2.0 < 3.0)
      MBrace.Core (>= 1.2 < 1.3)
      MBrace.Runtime (>= 1.2 < 1.3)
      WindowsAzure.ServiceBus (>= 2.5.2 < 2.6)
      WindowsAzure.Storage (>= 4.3 < 4.4)
    MBrace.Core (1.2.6)
    MBrace.Runtime (1.2.6)
      FsPickler (>= 2.1 < 2.2)
      FsPickler.Json (>= 2.1 < 2.2)
      MBrace.Core (1.2.6)
      Vagabond (>= 0.13 < 0.14)
    MBrace.Thespian (1.2.6)
      Argu (>= 2.0 < 3.0)
      FsPickler (>= 2.1 < 2.2)
      MBrace.Core (1.2.6)
      MBrace.Runtime (1.2.6)
      Thespian (>= 0.1.11-alpha < 0.2)
      Vagabond (>= 0.13 < 0.14)
    Microsoft.Data.Edm (5.7)
    Microsoft.Data.OData (5.7)
      Microsoft.Data.Edm (5.7)
      System.Spatial (5.7)
    Microsoft.Data.Services.Client (5.7)
      Microsoft.Data.OData (5.7)
    Microsoft.WindowsAzure.ConfigurationManager (3.2.1)
    Mono.Cecil (0.9.6.1)
    Newtonsoft.Json (8.0.3)
    System.Spatial (5.7)
    Thespian (0.1.11-alpha)
      FsPickler (>= 2.1 < 3.0)
    Vagabond (0.13)
      FsPickler (>= 2.1 < 3.0)
      Mono.Cecil (>= 0.9.6.1 < 0.9.7)
    WindowsAzure.ServiceBus (2.5.4)
      Microsoft.WindowsAzure.ConfigurationManager
    WindowsAzure.Storage (4.3)
      Microsoft.Data.OData (>= 5.6.2)
      Microsoft.Data.Services.Client (>= 5.6.2)
      Microsoft.WindowsAzure.ConfigurationManager (>= 1.8)
      Newtonsoft.Json (>= 5.0.8)
```

I've put it into a single file - same issue: -

```
#load @"packages\MBrace.Thespian\MBrace.Thespian.fsx"
#r @"packages\MBrace.Flow\lib\net45\MBrace.Flow.dll"
#r @"packages\Streams\lib\net45\Streams.dll"

open MBrace.Core
open MBrace.Flow
open MBrace.Library.Cloud
open MBrace.Thespian
open System

module Rop =
    type Result<'success, 'failure> = 
        | Success of 'success
        | Failure of 'failure

    /// Map a function over the success case
    let map f result = 
        match result with
        | Success success -> Success(f success)
        | Failure msgs -> Failure msgs
module Async =
    let map f xA = async { 
        let! x = xA
        return f x }
module AsyncResult =
    open Rop
    /// Lift a function to AsyncResult
    let map f (x:Async<Result<_,_>>) = Async.map (Rop.map f) x

let cluster = ThespianCluster.InitOnCurrentMachine 1

[| 1 .. 10 |]
|> CloudFlow.OfArray
|> CloudFlow.map(fun n -> n * 2)
|> CloudFlow.toArray
|> cluster.Run
```

**Comment by eiriktsarpalis** [comment]
I think I've found it. Looks like a Cecil bug https://github.com/jbevain/cecil/issues/278

**Comment by isaacabraham** [comment]
Sweet. Looks like there's already a fix in Cecil - if that's already been updated on NuGet I'll test it out.

Were you able to repro this?
